src/python/pants/backend/python/subsystems/python_native_code.py

```diff
@@ -99,26 +87,26 @@ def check_build_for_current_platform_only(self, targets):
     :return: a boolean value indicating whether the current target closure has native sources.
     :raises: :class:`pants.base.exceptions.IncompatiblePlatformsError`
     """
+    # TODO(#5949): convert this to checking if the closure of python requirements has any
+    # platform-specific packages (maybe find the platforms there too?).
     if not self._any_targets_have_native_sources(targets):
       return False
 
-    targets_by_platform = pex_build_util.targets_by_platform(targets, self._python_setup)
-    platforms_with_sources = self._get_targets_by_declared_platform_with_placeholders(targets_by_platform)
+    platforms_with_sources = pex_build_util.targets_by_platform(targets, self._python_setup)
     platform_names = list(platforms_with_sources.keys())
 
-    if len(platform_names) < 1:
-      raise self.PythonNativeCodeError(
-        "Error: there should be at least one platform in the target closure, because "
-        "we checked that there are native sources.")
-
-    if platform_names == ['current']:
+    if not platform_names:
+      return True
+    elif platform_names == ['current']:
       return True
 
-    raise IncompatiblePlatformsError(
-      'The target set contains one or more targets that depend on '
-      'native code. Please ensure that the platform arguments in all relevant targets and build '
-      'options are compatible with the current platform. Found targets for platforms: {}'
-      .format(str(platforms_with_sources)))
+    raise IncompatiblePlatformsError(dedent("""\
+      The target set contains one or more targets that depend on native code. Please ensure that the
```

Comment:
Let's be more specific in this error. What I'm hoping we can get to is roughly:

```
Pants doesn't currently support cross-compiling native code.
The following targets set platforms arguments other than ['current'], which is unsupported for this reason.
Please either remove the platforms argument from those targets, or set them to exactly ['current'].
Bad targets:
{}
```

Reply:
Also perhaps consider the recent change to remove this check being performed anywhere except in `PythonBinaryCreate` -- that might allow us to reduce the number of non-local exit paths in this method?